### DISCUSSION TITLE
Error messages at startup `TypeError: Cannot read property 'addClass' of null` with Settings Search plugin

### DISCUSSION THREAD

**Discussion by claremacrae** [Taitava/obsidian-shellcommands#233] [Debugging and testing]
Hello,

I noticed 2 errors in the console from this plugin at startup with:

- Obsidian 0.15.1 (Insider)
- Shell command: 0.12.1

Error message: `Uncaught TypeError: Cannot read property 'addClass' of null`

Output in the console:

```text
plugin:obsidian-shellcommands:7541 Uncaught TypeError: Cannot read property 'addClass' of null
    at HTMLButtonElement.tab_button_clicked (plugin:obsidian-shellcommands:7541)
    at createTabs (plugin:obsidian-shellcommands:7497)
    at SC_MainSettingsTab.display (plugin:obsidian-shellcommands:8521)
    at SettingsSearch.getTabResources (plugin:settings-search:170)
    at SettingsSearch.buildPluginResources (plugin:settings-search:120)
    at eval (plugin:settings-search:122)
tab_button_clicked @ plugin:obsidian-shellcommands:7541
createTabs @ plugin:obsidian-shellcommands:7497
display @ plugin:obsidian-shellcommands:8521
getTabResources @ plugin:settings-search:170
buildPluginResources @ plugin:settings-search:120
eval @ plugin:settings-search:122
setTimeout (async)
buildPluginResources @ plugin:settings-search:122
eval @ plugin:settings-search:122
setTimeout (async)
buildPluginResources @ plugin:settings-search:122
eval @ plugin:settings-search:122
setTimeout (async)
buildPluginResources @ plugin:settings-search:122
eval @ plugin:settings-search:122
setTimeout (async)
buildPluginResources @ plugin:settings-search:122
eval @ plugin:settings-search:122
setTimeout (async)
buildPluginResources @ plugin:settings-search:122
eval @ plugin:settings-search:122
setTimeout (async)
buildPluginResources @ plugin:settings-search:122
eval @ plugin:settings-search:122
setTimeout (async)
buildPluginResources @ plugin:settings-search:122
eval @ plugin:settings-search:122
setTimeout (async)
buildPluginResources @ plugin:settings-search:122
eval @ plugin:settings-search:122
setTimeout (async)
buildPluginResources @ plugin:settings-search:122
eval @ plugin:settings-search:122
setTimeout (async)
buildPluginResources @ plugin:settings-search:122
eval @ plugin:settings-search:122
setTimeout (async)
buildPluginResources @ plugin:settings-search:122
eval @ plugin:settings-search:122
setTimeout (async)
buildPluginResources @ plugin:settings-search:122
eval @ plugin:settings-search:105
(anonymous) @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
s @ app.js:1
Promise.then (async)
l @ app.js:1
s @ app.js:1
Promise.then (async)
l @ app.js:1
(anonymous) @ app.js:1
m @ app.js:1
t.loadLayout @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
s @ app.js:1
Promise.then (async)
l @ app.js:1
s @ app.js:1
Promise.then (async)
l @ app.js:1
s @ app.js:1
Promise.then (async)
l @ app.js:1
s @ app.js:1
Promise.then (async)
l @ app.js:1
s @ app.js:1
plugin:obsidian-shellcommands:7541 Uncaught TypeError: Cannot read property 'addClass' of null
    at HTMLButtonElement.tab_button_clicked (plugin:obsidian-shellcommands:7541)
    at SC_MainSettingsTab.rememberLastPosition (plugin:obsidian-shellcommands:8787)
    at SC_MainSettingsTab.display (plugin:obsidian-shellcommands:8576)
    at SettingsSearch.getTabResources (plugin:settings-search:170)
    at SettingsSearch.buildPluginResources (plugin:settings-search:120)
    at eval (plugin:settings-search:122)
```

**Comment by claremacrae** [comment]
Disabling the 'Settings Search' plugin made this error go away.

**Reply by Taitava** [reply]
Thanks for your report and also for this additional information!

I'll have to wait until Obsidian `0.15.x` is available to everybody before I can start examining this, if it's something specific to the new version of Obsidian. Do you know if _Settings Search_ and _Shell commands_ work together in older versions of Obsidian (`0.14.x`)? I haven't checked it yet, but might check later.

**Reply by claremacrae** [reply]
Here's the reply to the issue that I put in the Settings Search project:

https://github.com/valentine195/obsidian-settings-search/issues/17#issuecomment-1157524978
> Settings search renders settings tabs on load in order to grab settings elements. If a plugin has errors in their settings tabs, they will show up here.

**Reply by Taitava** [reply]
# I changed the title of the discussion
Old title: _Error messages at startup `TypeError: Cannot read property 'addClass' of null` with Obsidian 0.15.1_
New title: _Error messages at startup `TypeError: Cannot read property 'addClass' of null` with Settings Search plugin_

**Comment by Taitava** [comment]
## Inspecting the issue (still in progress)

I've used a couple of hours now to try to find out the reason for these errors. The root cause is still unclear to me.

https://github.com/Taitava/obsidian-shellcommands/blob/84602db18e994a64bb7a4e164b10b6b0bcd3f747/src/settings/setting_elements/Tabs.ts#L134-L135
134. This line searches for a tab content element in the application and assigns it to `tab_content`. For some reason, when Settings Search invokes the setting view to be rendered, this line cannot find the element, and so `tab_content` becomes `null`. **I need to inspect this part more**.
135. This line is when the symptom is noticed and an error message put to console log.

## AFAIK, the error messages do not prevent using the search or SC

Even though these error messages tell that something is going wrong under the hood, in my test I've seen that Settings Search is still able to find the Shell command plugin's settings, and SC is able to operate normally. In that sense, this bug is not a showstopper (= doesn't prevent usage). But who knows if it crashes badly in the future? That's why it would still be good to fix it.

@claremacrae have you noticed any actual, practical issues when using these two plugins together? Of course error messages in the console log should be fixed, and I'm in no way trying to convince you to ignore them. I just want to understand the nature of the problem better.

## Other glitches between SC and Settings Search

When I'm trying to fix compatibility issues with other plugins, I also need to consider what level of compatibility can realistically be achieved. Even if I get this particular error fixed, there are a couple of things that will not work between SC and Settings Search:

- If a setting is located in a secondary tab (e.g. tab _Environments_), Settings Search is not able to show that setting when clicking an arrow button. This is because the tab UI is specifically developed for this plugin by me, so it's not a standard feature in Obsidian. It's understandable that Settings Search cannot handle this kind of custom tab system. Settings Search can **find** settings from all tabs, but it cannot redirect users to go to the actual place where the setting resides.
- When a new shell command is created, Settings Search does not find it. Obsidian needs to be restarted for Settings Search to notice the new  shell command.
- Search Settings cannot find any settings contained in modals. I think this is perfectly ok and probably no-one would even need to find those settings with Search Settings, as all modal settings are repetitive and would probably just flood the search results if a user has tens of shell commands defined.

I can see that even with those glitches, it does make sense to use Settings Search together with SC, even if the search is not able to fully support all settings of SC. But it might cause some confusion among users, and more reports about incompatibility.
